front/FrontService.cpp

```diff
@@ -19,5 +19,478 @@
+    onSendMessage(_moduleID, _nodeID, uuid, _data);
+
+  } catch (std::exception &e) {
+    FRONT_LOG(ERROR) << LOG_BADGE("asyncSendMessageByNodeID")
+                     << LOG_KV("what", boost::diagnostic_information(e));
+  }
+}
+
+/**
+ * @brief: send message to multiple nodes
+ * @param _moduleID: moduleID
+ * @param _nodeIDs: the receiver nodeIDs
+ * @param _data: send data
+ * @return void
+ */
+void FrontService::asyncSendMessageByNodeIDs(int _moduleID,
+                                             const crypto::NodeIDs &_nodeIDs,
+                                             bytesConstRef _data) {
+  for (const auto &_nodeID : _nodeIDs) {
+    asyncSendMessageByNodeID(_moduleID, _nodeID, _data, 0, CallbackFunc());
+  }
+}
+
+/**
+ * @brief: send broadcast message
+ * @param _moduleID: moduleID
+ * @param _data:  send data
+ * @return void
+ */
+void FrontService::asyncMulticastMessage(int _moduleID, bytesConstRef _data) {
+
+  auto message = messageFactory()->buildMessage();
+  message->setModuleID(_moduleID);
+  message->setPayload(_data);
+
+  auto buffer = std::make_shared<bytes>();
+  message->encode(*buffer.get());
+
+  m_gatewayInterface->asyncMulticastMessage(
+      m_groupID, bytesConstRef(buffer->data(), buffer->size()));
+}
+
+/**
+ * @brief: register the node change callback
+ * @param _moduleID: moduleID
+ * @param _notifier: callback
+ * @return void
+ */
+void FrontService::registerNodeStatusNotifier(int _moduleID,
+                                              NodeStatusNotifier _notifier) {
+  m_mapNodeStatusNotifier[_moduleID] = _notifier;
+
+  FRONT_LOG(INFO) << LOG_BADGE("registerNodeStatusNotifier")
+                  << LOG_KV("moduleID", _moduleID)
+                  << LOG_KV("nodeID", m_nodeID->hex())
+                  << LOG_KV("groupID", m_groupID);
+}
+
+/**
+ * @brief: register the callback for module message
+ * @param _moduleID: moduleID
+ * @param _dispatcher: callback
+ * @return void
+ */
+void FrontService::registerMessageDispatcher(int _moduleID,
+                                             MessageDispatcher _dispatcher) {
+  m_mapMessageDispatcher[_moduleID] = _dispatcher;
+
+  FRONT_LOG(INFO) << LOG_BADGE("registerMessageDispatcher")
+                  << LOG_KV("moduleID", _moduleID)
+                  << LOG_KV("nodeID", m_nodeID->hex())
+                  << LOG_KV("groupID", m_groupID);
+}
+
+/**
+ * @brief: receive nodeIDs from gateway
+ * @param _error: error info
+ * @param _nodeIDs: received nodeIDs
+ * @return void
+ */
+void FrontService::onReceiveNodeIDs(
+    Error::Ptr _error, std::shared_ptr<const bcos::crypto::NodeIDs> _nodeIDs) {
+
+  // error
+  if (_error && _error->errorCode() != CommonError::SUCCESS) {
+    FRONT_LOG(ERROR) << LOG_BADGE("onReceiveNodeIDs")
+                     << LOG_KV("errorCode", _error->errorCode())
+                     << LOG_KV("errorMessage", _error->errorMessage());
+    return;
+  }
+
+  {
+    RecursiveGuard l(x_nodeIDs);
+    m_nodeIDs = _nodeIDs;
```

Comment:
assign the old node list directly with a recursive guard?

Reply:
instead of RecursiveMutex with Mutex